doc/wavetable-communication.md

```diff
@@ -0,0 +1,184 @@
+  o...ooo 2D Tensor (Tensor1s for each freq)      | or just
+  ^       with ringbuffer semantics (see below)   | "waves"
+  |       (MW produces, AD note consumes)         |
+  |                                               |
+  o...ooo 3D Tensor (Tensor2s for each semantic) --
+
+```
+
+Ringbuffer layout:
+
+```
+  no required space holder (r=w is unambiguous internally)             wrap
+         |                                                               |
+         v               |--reserved for write--|                        v
+  ...----|--can be read--|                      |--can be reserved next...
+  ...oooooooo............oooo...................oooo......................
+         |               |                      |
+         r               w_delayed              w              <- pointers
+```
+
+Wavetable layout:
+
+```
+ o scales for ringbuffer (Tensor1, semantics + freqs)
+ o ringbuffer (Tensor3, wave data)
+
+ o timestamp of params for current ringbuffer
+ o timestamp of params for next ("to be filled") ringbuffer
+
+ o wavetable mode (e.g. random seeds, basefunc parameter, ...)
+```
+
+ADnote contains two wavetables per OscilGen:
+
+- current wavetable (in used, can be refilled with new waves)
+- next wavetable (under construction after parameter change)
+
+## Asynchonicity principles
+
+MiddleWare and ADnoteParams are run by 2 different threads at the same
+time. This can cause issues, like outdated requests. To avoid this, the
+following principles have been set up:
+
+* wavetable requests induced by parameter changes always carry a timestamp
+* any side should not rely on the other side handling the asynchonicity
+  "well", i.e. it can not be relied on the other side detectig messages
+  with outdated timestamps
+
+## Sequence diagram
+
+The following (not necessarily UML standard) diagram explains
+how and when new wavetables are being generated. There are two
+entry points, marked with an "X".
+
+```
+    MW (non-RT)                                                 ADnote (RT)
+    |                                                           |
+    |   wavetable-relevant params changed                       |
+    |<---X                                                      |
+    |                                                           |
+    |  If "wavetable-params-changed" is not suppressed:         |
+    |--/path/to/advoice/wavetable-params-changed:Ti:Fi--------->|
+    |      Inform ADnote that data relevant for wavetable       |
+    |      creation has changed                                 |
+    |      - path: osc or mod-osc? (T/F)                        |
+    |      - unique timestamp of parameter change (i)           |
+    |                                                           |
+    - suppress further "wavetable-params-changed"               |
+    |                                                           |
+    |                          - mark current ringbuffer        |
```

Comment:
One ringbuffer per target location (part/kit/add/voice/osc-mod|osc), right?

Reply:
Note: This was an error in the description. It meant: "mark current wavetable as outdated". There is not "current ringbuffer". Will be fixed with next commit.